core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -101,122 +101,134 @@ public void configure() {
+         ClusteredGetCommand getCommand, RpcOptionsBuilder rpcOptionsBuilder, int lastTopologyId) {
+      if (value != null)
+         return CompletableFuture.completedFuture(value);
+
+      final CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
+      final int currentTopologyId = cacheTopology.getTopologyId();
+
+      if (trace) {
             log.tracef("Perform remote get for key %s. topologyId=%s, currentTopologyId=%s",
                        key, lastTopologyId, currentTopologyId);
-         }
-         List<Address> targets;
-         if (lastTopologyId < currentTopologyId) {
-            // Cache topology has changed or it is the first time.
-            lastTopologyId = currentTopologyId;
-            targets = new ArrayList<>(cacheTopology.getReadConsistentHash().locateOwners(key));
-         } else if (lastTopologyId == currentTopologyId && cacheTopology.getPendingCH() != null) {
-            // Same topologyId, but the owners could have already installed the next topology
-            // Lets try with pending consistent owners (the read owners in the next topology)
-            lastTopologyId = currentTopologyId + 1;
-            targets = new ArrayList<>(cacheTopology.getPendingCH().locateOwners(key));
-            // Remove already contacted nodes
-            targets.removeAll(cacheTopology.getReadConsistentHash().locateOwners(key));
-            if (targets.isEmpty()) {
-               if (trace) {
-                  log.tracef("No valid values found for key '%s' (topologyId=%s).", key, currentTopologyId);
-               }
-               break;
-            }
-         } else { // lastTopologyId > currentTopologyId || cacheTopology.getPendingCH() == null
-            // We have not received a valid value from the pending CH owners either, and the topology id hasn't changed
+      }
+      List<Address> targets;
+      int newTopologyId;
+      if (lastTopologyId < currentTopologyId) {
+         // Cache topology has changed or it is the first time.
+         newTopologyId = currentTopologyId;
+         targets = new ArrayList<>(cacheTopology.getReadConsistentHash().locateOwners(key));
+      } else if (lastTopologyId == currentTopologyId && cacheTopology.getPendingCH() != null) {
+         // Same topologyId, but the owners could have already installed the next topology
+         // Lets try with pending consistent owners (the read owners in the next topology)
+         newTopologyId = currentTopologyId + 1;
+         targets = new ArrayList<>(cacheTopology.getPendingCH().locateOwners(key));
+         // Remove already contacted nodes
+         targets.removeAll(cacheTopology.getReadConsistentHash().locateOwners(key));
+         if (targets.isEmpty()) {
             if (trace) {
                log.tracef("No valid values found for key '%s' (topologyId=%s).", key, currentTopologyId);
             }
-            break;
+            return CompletableFuture.completedFuture(null);
```

Comment:
maybe we can cache the null completed future in ```CompletableFutures```? it seems to be used everywhere :P